### PR TITLE
probe: Add test for guest agent ping liveliness probe

### PR DESCRIPTION
**Pull request by akrejcir**
**What this PR does / why we need it**:
The guest agent liveliness probe was not tested before.

**Release note**:
```release-note
None
```

### CODE REVIEW

File: tests/network/probes.go
Comment (maiqueb) [inline tests/network/probes.go:245]: why do you need to pass `gomega` as input ? afaiu, that pkg is imported without a qualifier, meaning it will be globally available in the file.

File: tests/network/probes.go
Comment (maiqueb) [inline tests/network/probes.go:236]: nit: could we inline these calls ? 

feel free to disregard.

File: tests/network/probes.go
Comment (maiqueb) [inline tests/network/probes.go:243]: nit: I would move this to the test setup block, leaving only the expectation (eventually) in the test's body.

File: tests/network/probes.go
Comment (maiqueb) [reply tests/network/probes.go:245]: This reads weird, especially the part about `...Should(Succeed())` since your function does not return anything.

... also not fond of embedding `Expect`s within an `Eventually` block.

I would re-write that helper func to something like this:
```golang
Eventually(func(g Gomega) (*VirtualMachineInstance, error){
    foundVmi, err := virtClient.VirtualMachineInstance(util.NamespaceTestDefault).Get(vmi.Name, &metav1.GetOptions{})
    if err != nil {
        return nil, err
    }
    if !foundVmi.IsFinal() {
        return nil, fmt.Errorf("VMI %s still running", foundVmi.GetName())
    }
}).Should(BeTrue(), ...)
```

**Review by maiqueb — CHANGES_REQUESTED** [review]
Looks good - except the function you're passing to the eventually block.

File: tests/network/probes.go
Comment (akrejcir) [reply tests/network/probes.go:245]: It is easier to pass a `Gomega` instance to the function in `Eventually`. Then you can call `g.Expect()` on it, and it will not fail the whole test, only the current eventually loop and then it will try again later. 
Here is the documentation, check the few paragraphs at the end.
https://pkg.go.dev/github.com/onsi/gomega#Eventually

Basically this is a special syntax, if you want to check multiple thinks inside the loop.

Comment (EdDev) [reply tests/network/probes.go:245]: That documentation also mentions “complex systems”, which I interpret as complex tests.
This one is not complex IMO.
There are two disadvantages with this approach:
- It is not commonly used in the code base and is somehow the opposite pattern.
   The existing usages use assertions in Eventually to fail earlier and avoid retry.
- The assertions are mixed into two levels, in the body and outside. Instead, the body can just perform the functionality and allow the Eventually matcher to do the assertion.

Comment (EdDev) [reply tests/network/probes.go:245]: Just to clarify, this is equivalent (and IMO simpler):

```go
Eventually(func() (*v1.VirtualMachineInstance, error) {
    return virtClient.VirtualMachineInstance(vmi.Namespace).Get(vmi.Name, &metav1.GetOptions{})
}, 2*time.Minute, 1*time.Second).Should(Or(BeInPhase(v1.Failed), HaveSucceeded()))
```

This version has the advantage of showing the actually phase and not just a boolean.

Comment (akrejcir) [reply tests/network/probes.go:245]: Ok, I understand.
I didn't know about matcher `BeInPhase()`, thanks. Using it is of course simpler, I will change the code.

File: tests/network/probes.go
Comment (akrejcir) [reply tests/network/probes.go:243]: I think it is part of the test, not setup. The test checks that if guest agent is disabled, then the VM will stop.

File: tests/network/probes.go
Comment (akrejcir) [reply tests/network/probes.go:236]: I want to keep the vmi object creation and adding it to the cluster separate.

File: tests/network/probes.go
Comment (maiqueb) [reply tests/network/probes.go:243]: My point exactly: if the test checks that when the guest agent is disabled, the VM will stop, the test should consist exactly of that expectation.

Anyway, this is just a nit - you can ignore it.

File: tests/network/probes.go
Comment (maiqueb) [reply tests/network/probes.go:236]: I agree that might be useful when debugging - something I would accept for a unit test.

Anyway, your call - I won't block over this.

Comment (maiqueb) [reply tests/network/probes.go:237]: Do we really need `VMILauncherIgnoreWarnings` ? Can't we use the regular func to launcher the VMI ?

**Review by maiqueb — COMMENTED** [review]
Just a question.

File: tests/network/probes.go
Comment (akrejcir) [reply tests/network/probes.go]: I copied the code from other test.
Ok, I've changed it to `tests.RunVMIAndExpectLaunchIgnoreWarnings()`.

Comment (akrejcir) [reply tests/network/probes.go]: Ok, I will change it. I don't know why the other tests ignore warnings.

Comment (maiqueb) [reply tests/network/probes.go]: Thank you; please let us know what warnings are thrown if this happens to fail, and we can evaluate if those warnings are reasonable or not.

File: tests/network/probes.go
Comment (lyarwood) [inline tests/network/probes.go:248]: This will fail errcheck once this file is whitelisted FWIW so we may as well capture and check `err` here.

File: tests/network/probes.go
Comment (lyarwood) [inline tests/network/probes.go:249]: Huh, unrelated as this is my first time really looking at the behaviour of Liveness probes but does it really leave the VMI in a `v1.Failed` condition?

Comment (lyarwood) [reply tests/network/probes.go:249]: Answering own question yes it does, apologies the noise!

**Comment by akrejcir** [comment]
@phoracek , can you please take a look and approve, if it is ok? Or do you know who could approve this?

**Review by maiqueb — APPROVED** [review]
I can't approve, but looks good.

Thanks for taking the time to attempt this without ignoring errors - sorry for the added time it took.

**Comment by akrejcir** [comment]
/cc @davidvossel @vladikr 
I'm not sure who can approve this. Can you please take a look, or mention someone who can approve?

**Review by davidvossel — COMMENTED** [review]
/approve

**Pull request merged**
